numba/np/arrayobj.py

```diff
@@ -1339,6 +1339,26 @@ def codegen(context, builder, sig, args):
     return sig, codegen
 
 
+@intrinsic
+def get_readonly_array(typingctx, arr):
+    # returns a copy of arr which is readonly
+    ret = arr.copy(readonly=True)
+    sig = ret(arr)
+
+    def codegen(context, builder, sig, args):
+        [src] = args
+        srcty = sig.args[0]
+
+        dest = make_array(srcty)(context, builder, src)
+        # Hack to return a read-only array
+        setattr(dest, 'parent', Constant(
+                context.get_value_type(dest._datamodel.get_type('parent')),
+                None))
```

Comment:
I have a question here. What's the difference between `parent` as an attribute and `parent` as a type within the datamodel ?

Reply:
`parent` as an attribute is a member of the ndarray struct, whereas the datamodel says how `parent` should be typed. See:
https://github.com/numba/numba/blob/2e9b58cfce0391b968ab8c7c84393cc12c4d1bfb/numba/core/datamodel/models.py#L876